src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/ApplicationsController.kt

```diff
@@ -55,10 +56,34 @@ class ApplicationsController(
       is ValidatableActionResult.GeneralValidationError -> throw BadRequestProblem(errorDetail = applicationResult.message)
       is ValidatableActionResult.FieldValidationError -> throw BadRequestProblem(invalidParams = applicationResult.validationMessages)
       is ValidatableActionResult.Success -> applicationResult.entity
+      else -> shouldNotBeReached()
     }
 
     return ResponseEntity
       .created(URI.create("/applications/${application.id}"))
       .body(applicationsTransformer.transformJpaToApi(application))
   }
+
+  override fun applicationsApplicationIdPut(applicationId: UUID, body: UpdateApplication): ResponseEntity<Application> {
+    val deliusPrincipal = httpAuthService.getDeliusPrincipalOrThrow()
+    val username = deliusPrincipal.name
+
+    val serializedData = objectMapper.writeValueAsString(body.data)
+
+    val validationResult = when (val applicationResult = applicationService.updateApplication(applicationId, serializedData, body.submittedAt, username)) {
+      is AuthorisableActionResult.NotFound -> throw NotFoundProblem(applicationId, "Application")
+      is AuthorisableActionResult.Unauthorised -> throw ForbiddenProblem()
+      is AuthorisableActionResult.Success -> applicationResult.entity
+      else -> shouldNotBeReached()
+    }
+
+    val updatedApplication = when (validationResult) {
+      is ValidatableActionResult.GeneralValidationError -> throw BadRequestProblem(errorDetail = validationResult.message)
+      is ValidatableActionResult.FieldValidationError -> throw BadRequestProblem(invalidParams = validationResult.validationMessages)
+      is ValidatableActionResult.Success -> validationResult.entity
+      else -> shouldNotBeReached()
+    }
```

Comment:
I find this a bit tricky to follow. I think it might be easier if we remove the setting of the `applicationResult` value from the conditional, so:

```kotlin
val applicationResult = applicationService.updateApplication(applicationId, serializedData, body.submittedAt, username)

val validationResult = when (applicationResult) {
  is AuthorisableActionResult.NotFound -> throw NotFoundProblem(applicationId, "Application")
  is AuthorisableActionResult.Unauthorised -> throw ForbiddenProblem()
  is AuthorisableActionResult.Success -> applicationResult.entity
  else -> shouldNotBeReached()   
}

val updatedApplication = when (validationResult) {
  is ValidatableActionResult.GeneralValidationError -> throw BadRequestProblem(errorDetail = validationResult.message)
  is ValidatableActionResult.FieldValidationError -> throw BadRequestProblem(invalidParams = validationResult.validationMessages)
  is ValidatableActionResult.Success -> validationResult.entity
  else -> shouldNotBeReached()
}
```

Would this work?